packages/lib-ripple/index.ts

```diff
@@ -0,0 +1,68 @@
+import { DfnsApiClient, DfnsError } from '@dfns/sdk'
+import { Transaction, deriveAddress, decode, encode } from 'xrpl'
+import { GetWalletResponse, GenerateSignatureResponse } from '@dfns/sdk/types/wallets'
+
+export type DfnsWalletOptions = {
+  walletId: string
+  dfnsClient: DfnsApiClient
+}
+
+
+const assertSigned = (res: GenerateSignatureResponse) => {
+  if (res.status === 'Failed') {
+    throw new DfnsError(-1, 'signing failed', res)
+  } else if (res.status !== 'Signed') {
+    throw new DfnsError(
+      -1,
+      'cannot complete signing synchronously because this wallet action requires policy approval',
+      res
+    )
+  }
+}
+
+type WalletMetadata = GetWalletResponse
+
+export class DfnsWallet {
```

Comment:
looking into how to make the integration tighter, this is what we should do.

xrpl offers a [wallet](https://github.com/XRPLF/xrpl.js/blob/main/packages/xrpl/src/Wallet/index.ts) abstraction. unfortunately it's not an interface. but we can still make this work by extending the class.

```
export class DfnsWallet extends xrpl.Wallet
```

we will override the following implementations

```
public get address(): string
public sign(transaction: Transaction, multisign?: boolean | string): {
    tx_blob: string
    hash: string
}
public getXAddress(tag: number | false = false, testnet = false): string
```

should also override this, but throw not implemented error

```
public verifyTransaction(): boolean
```

nice thing with this design is then we can integrate with the xrpl much better like this

```
const res = await client.submitAndWait({
    TransactionType: 'Payment',
    Account: wallet.address,
    Destination: "rBYtCQKxGTfFuob3hxSc8pEYddetT9CdDZ",
    Amount: '1',
  },
  {
    autofill: true,
    wallet,
  })
```

Reply:
nevermind, this would've worked except I just realized their `sign` function is not async. even though if they just made it async, it would work. `submitAndWait` is async.

that said, would still make sense I think to match the return type of `sign` to their function. minimize the differences between the examples they have and our example.

```
Promise<{
    tx_blob: string
    hash: string
}>
```

Reply:
Yeah, I've seen their wallet abstraction, I did't think that it will be worth it. After seeing you're example, maybe it would have been more straight forward for the user. But unfortunately we can't use it :) 

I'll change the "sign" return. I'll keep in mind that we always want to keep things as close to the official library as possible.